File: packages/hushspec/tests/panic.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import {
  evaluate,
  activatePanic,
  deactivatePanic,
  isPanicActive,
  panicPolicy,
} from '../src/evaluate.js';
import { parseOrThrow } from '../src/parse.js';
import type { HushSpec } from '../src/schema.js';

describe('panic mode', () => {
  beforeEach(() => {
    deactivatePanic();
  });

  afterEach(() => {
    deactivatePanic();
  });

  it('isPanicActive returns false by default', () => {
    expect(isPanicActive()).toBe(false);
  });

  it('activatePanic sets panic active', () => {
    activatePanic();
    expect(isPanicActive()).toBe(true);
  });

  it('deactivatePanic clears panic active', () => {
    activatePanic();
    expect(isPanicActive()).toBe(true);
    deactivatePanic();
    expect(isPanicActive()).toBe(false);
  });

  it('evaluate returns deny for all action types during panic', () => {
    activatePanic();

    const spec: HushSpec = { hushspec: '0.1.0' };
    const actionTypes = [
      'tool_call',
      'egress',
      'file_read',
      'file_write',
      'patch_apply',
      'shell_command',
      'computer_use',
      'unknown_action',
    ];

    for (const actionType of actionTypes) {
      const result = evaluate(spec, { type: actionType, target: 'anything' });
      expect(result.decision).toBe('deny');
      expect(result.matched_rule).toBe('__hushspec_panic__');
      expect(result.reason).toBe('emergency panic mode is active');
    }
  });

  it('deactivate restores normal evaluation', () => {
    const spec: HushSpec = { hushspec: '0.1.0' };
    const action = { type: 'tool_call', target: 'some_tool' };

    let result = evaluate(spec, action);
    expect(result.decision).toBe('allow');

    activatePanic();
    result = evaluate(spec, action);
    expect(result.decision).toBe('deny');

    deactivatePanic();
    result = evaluate(spec, action);
    expect(result.decision).toBe('allow');
  });

  it('panicPolicy returns a valid HushSpec', () => {
    const spec = panicPolicy();
    expect(spec.hushspec).toBe('0.1.0');
    expect(spec.name).toBe('__hushspec_panic__');
    expect(spec.rules).toBeDefined();
    expect(spec.rules!.forbidden_paths).toBeDefined();
    expect(spec.rules!.egress).toBeDefined();
    expect(spec.rules!.tool_access).toBeDefined();
    expect(spec.rules!.shell_commands).toBeDefined();
    expect(spec.rules!.computer_use).toBeDefined();
  });

  it('panic policy denies file reads', () => {
    const spec = panicPolicy();
    const result = evaluate(spec, { type: 'file_read', target: '/etc/passwd' });
    expect(result.decision).toBe('deny');
  });

  it('panic policy denies egress', () => {
    const spec = panicPolicy();
    const result = evaluate(spec, { type: 'egress', target: 'example.com' });
    expect(result.decision).toBe('deny');
  });

  it('panic policy denies tool calls', () => {
    const spec = panicPolicy();
    const result = evaluate(spec, { type: 'tool_call', target: 'any_tool' });
    expect(result.decision).toBe('deny');
  });
});

File: packages/go/hushspec/panic_test.go
package hushspec

import (
	"os"
	"path/filepath"
	"testing"
)

func resetPanic() {
	DeactivatePanic()
}

func TestPanicActivateDeactivate(t *testing.T) {
	resetPanic()
	defer resetPanic()

	if IsPanicActive() {
		t.Fatal("expected panic to be inactive initially")
	}

	ActivatePanic()
	if !IsPanicActive() {
		t.Fatal("expected panic to be active after ActivatePanic()")
	}

	DeactivatePanic()
	if IsPanicActive() {
		t.Fatal("expected panic to be inactive after DeactivatePanic()")
	}
}

func TestPanicModeDeniesAll(t *testing.T) {
	resetPanic()
	defer resetPanic()

	ActivatePanic()

	spec := &HushSpec{HushSpecVersion: "0.1.0"}

	actionTypes := []string{
		"tool_call", "egress", "file_read", "file_write",
		"patch_apply", "shell_command", "computer_use", "unknown_action",
	}

	for _, actionType := range actionTypes {
		action := &EvaluationAction{
			Type:   actionType,
			Target: "anything",
		}
		result := Evaluate(spec, action)
		if result.Decision != DecisionDeny {
			t.Errorf("expected deny for action type %q, got %q", actionType, result.Decision)
		}
		if result.MatchedRule != "__hushspec_panic__" {
			t.Errorf("expected matched_rule '__hushspec_panic__', got %q", result.MatchedRule)
		}
		if result.Reason != "emergency panic mode is active" {
			t.Errorf("expected panic reason, got %q", result.Reason)
		}
	}
}

func TestDeactivateRestoresNormal(t *testing.T) {
	resetPanic()
	defer resetPanic()

	spec := &HushSpec{HushSpecVersion: "0.1.0"}
	action := &EvaluationAction{
		Type:   "tool_call",
		Target: "some_tool",
	}

	result := Evaluate(spec, action)
	if result.Decision != DecisionAllow {
		t.Fatalf("expected allow in normal mode, got %q", result.Decision)
	}

	ActivatePanic()
	result = Evaluate(spec, action)
	if result.Decision != DecisionDeny {
		t.Fatalf("expected deny in panic mode, got %q", result.Decision)
	}

	DeactivatePanic()
	result = Evaluate(spec, action)
	if result.Decision != DecisionAllow {
		t.Fatalf("expected allow after deactivation, got %q", result.Decision)
	}
}

func TestSentinelFileActivatesPanic(t *testing.T) {
	resetPanic()
	defer resetPanic()

	dir := t.TempDir()
	sentinel := filepath.Join(dir, ".hushspec_panic")

	if err := os.WriteFile(sentinel, []byte(""), 0644); err != nil {
		t.Fatal(err)
	}

	if !CheckPanicSentinel(sentinel) {
		t.Fatal("expected CheckPanicSentinel to return true when file exists")
	}
	if !IsPanicActive() {
		t.Fatal("expected panic to be active after sentinel check")
	}
}

func TestSentinelFileMissingDoesNotActivate(t *testing.T) {
	resetPanic()
	defer resetPanic()

	sentinel := filepath.Join(t.TempDir(), "nonexistent")

	if CheckPanicSentinel(sentinel) {
		t.Fatal("expected CheckPanicSentinel to return false when file missing")
	}
	if IsPanicActive() {
		t.Fatal("expected panic to remain inactive when sentinel missing")
	}
}
